### PR TITLE
parser: support sql_mode 'IGNORE SPACE'

### PR DESCRIPTION
**Pull request by spongedu**
support sql_mode 'IGNORE SPACE' for TiDB, and refine function name resolution rule according to MySQL's behavior. See https://dev.mysql.com/doc/refman/5.7/en/function-resolution.html for details.

### CODE REVIEW

**Comment by shenli** [comment]
@spongedu Thanks!
Please resolve the conflicts.

File: mysql/const.go [hunk 1]
```diff
@@ -452,6 +452,16 @@ func (m SQLMode) HasRealAsFloatMode() bool {
 	return m&ModeRealAsFloat == ModeRealAsFloat
 }
 
+// HasNoBackslashEscapesMode detects if 'NO_BACKSLASH_ESCAPES' mode is set in SQLMode
```
Comment (shenli) [inline mysql/const.go]: This function is not related with IGNORE_SPACE.

File: parser/parser_test.go [hunk 2]
```diff
@@ -828,7 +828,7 @@ func (s *testParserSuite) TestBuiltin(c *C) {
 
 		// for date, day, weekday
 		{"SELECT CURRENT_DATE, CURRENT_DATE(), CURDATE()", true},
-		{"SELECT CURRENT_DATE, CURRENT_DATE(), CURDATE(1)", true},
+		{"SELECT CURRENT_DATE, CURRENT_DATE(), CURDATE(1)", false},
```
Comment (shenli) [inline parser/parser_test.go]: Why this is false case?

File: mysql/const.go (hunk 1 above)
Comment (spongedu) [reply mysql/const.go]: @shenli  It seems to be taken into this pr in a merge. I'll delete this.

File: parser/parser_test.go (hunk 2 above)
Comment (spongedu) [reply parser/parser_test.go]: @shenli  it seems that MySQL doesn't support `CURDATE` with one parameter. see: https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_curdate

**Comment by zz-jason** [comment]
@spongedu plz resolve conflicts.

File: parser/parser.y [hunk 3]
```diff
@@ -2989,10 +3019,18 @@ FunctionCallKeyword:
 	{
 		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: $3.([]ast.ExprNode)}
 	}
+|   builtinUser '('	ExpressionListOpt ')'
```
Comment (zz-jason) [inline parser/parser.y]: keep the code indentation

File: parser/lexer.go [hunk 4]
```diff
@@ -435,7 +470,16 @@ func scanIdentifier(s *Scanner) (int, Pos, string) {
 	pos := s.r.pos()
 	s.r.inc()
 	s.r.incAsLongAs(isIdentChar)
-	return identifier, pos, s.r.data(&pos)
+	lit := s.r.data(&pos)
```
Comment (tiancaiamao) [inline parser/lexer.go]: It's quite performance sensitive here.
`strings.ToLower(lit)` would cause allocation, and map access is not that fast.
I think you can switch the order, check IgnoreSpaceMode first, try the best not to hurt the performance in most cases.

File: parser/parser.y (hunk 3 above)
Comment (spongedu) [reply parser/parser.y]: @zz-jason ok

File: parser/lexer.go (hunk 4 above)
Comment (spongedu) [reply parser/lexer.go]: @tiancaiamao ok, I'll take a look. thx :)

File: parser/lexer.go [hunk 5]
```diff
@@ -25,6 +25,41 @@ import (
 
 var _ = yyLexer(&Scanner{})
 
+var ignoreSpaceFuncMap = map[string]int{
```
Comment (zz-jason) [inline parser/lexer.go]: How do we obtain tis map ?

File: parser/lexer.go (hunk 4 above)
Comment (spongedu) [reply parser/lexer.go]: @tiancaiamao I've refine this place. PTAL

File: parser/lexer.go (hunk 5 above)
Comment (spongedu) [reply parser/lexer.go]: @zz-jason I've rewrite this part. PTAL

**Comment by tiancaiamao** [comment]
LGTM

**Comment by tiancaiamao** [comment]
PTAL @zz-jason @shenli @XuHuaiyu

File: parser/parser.y [hunk 6]
```diff
@@ -3123,10 +3149,18 @@ FunctionCallKeyword:
...
 	}
+|	builtinUser '('	ExpressionListOpt ')'
+	{
+		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: $3.([]ast.ExprNode)}
+	}
 |	FunctionNameOptionalBraces OptionalBraces
 	{
 		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1)}
 	}
+|   builtinCurDate '(' ')'
```
[the remark this answers is not in the document]
Comment (spongedu) [reply parser/parser.y]: ok

File: parser/parser.y [hunk 7]
```diff
@@ -3182,7 +3216,11 @@ FunctionCallKeyword:
 	}
 
 FunctionCallNonKeyword:
-	"CURTIME" '(' FuncDatetimePrecListOpt ')'
+	builtinCurTime '(' FuncDatetimePrecListOpt ')'
+	{
+		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: $3.([]ast.ExprNode)}
+	}
+| builtinSysDate '(' FuncDatetimePrecListOpt ')'
```
[the remark this answers is not in the document]
Comment (spongedu) [reply parser/parser.y]: done

**Comment by spongedu** [comment]
@zimulala Fixed the failed cases. PTAL, Thanks :)
